docs/proposals/scheduling/20220530-fine-grained-cpu-orchestration.md

```diff
@@ -0,0 +1,690 @@
+    1. koordlet ignore the CPUs reserved by kubelet, and replace them with CPU Shared Pool defined by Koordinator. 
+1. For K8s BestEffort and Koordinator BE Pods:
+    1. If kubelet reserved CPUs, the best effort Pods use the reserved CPUs first.
+    1. koordlet can use all CPUs in the node but exclude the CPUs allocated by K8s Guaranteed and Koordinator LSE Pods that have integer CPU. It means that if koordlet enables the CPU Suppress feature should follow the constraint to guarantee not affecting LSE Pods. Similarly, if kubelet enables the CPU manager policy with `static`, the K8s Guaranteed Pods should also be excluded. 
+1. For K8s Guaranteed Pod CPU:
+    1. If there is `scheduler.koordinator.sh/alloc-status` updated by koord-scheduler in the Pod annotation, then replace the CPUSet in the kubelet CRI request, including Sandbox/Container creating stage.
+    1. kubelet sometimes call `Update` method defined in CRI to update container cgroup to set new CPUs, so koordlet and koord-runtime-proxy need to hook the method.
+1. Automatically resize CPU Shared Pool
+    1. koordlet automatically resize `CPU shared pool` based on the changes such as Pod creating/destroying. If `CPU shared pool` changed, koordlet should update cgroups of all LS/K8s Burstable Pods with the CPUs of shared pool. 
+    1. If the corresponding CPU Shared Pool is specified in the annotation `scheduler.koordlet.sh/alloc-status` of the Pod, koordlet need to bind only the CPUs of the corresponding pool when configuring the cgroup.
+
+The takeover logic will require koord-runtime-proxy to add new extension points, and require koordlet to implement a new runtime hook plugin. When koord-runtime-proxy is not installed, these takeover logic will also be able to be implemented.
+
+### CPU orchestration API
+
+#### Application CPU CPU orchestration API
+
+##### Resource request spec
+
+The annotation `scheduler.koordinator.sh/alloc-spec` is a resource allocation API defined by Koordinator. The user specifies the desired CPU orchestration strategy by setting this annotation. In the future, we can also extend and add resource types that need to be supported as needed. The scheme corresponding to the annotation value is defined as follows:
+
+```go
+// ResourceAllocSpec describes extra attributes of the compute resource requirements.
+type ResourceAllocSpec struct {
+  CPUBindStrategy    CPUBindStrategy    `json:"cpuBindStrategy,omitempty"`
+  CPUIsolateStrategy CPUIsolateStrategy `json:"cpuIsolateStrategy,omitempty"`
+}
+
+type CPUBindStrategy string
+
+const (
+  // CPUBindStrategySameCore favor cpuset allocation that pack in few physical cores
+  CPUBindStrategySameCore CPUBindStrategy = "SameCore"
+  // CPUBindStrategySpread favor cpuset allocation that evenly allocate logical cores across physical cores
+  CPUBindStrategySpread   CPUBindStrategy = "Spread"
+)
+
+type CPUIsolateStrategy string
```

Comment:
CPUExclusivePolicy？

available values ares pcpu/numa-node

Reply:
Consider hiding this field first and adding it if necessary

Reply:
> CPUExclusivePolicy？
> 
> available values ares pcpu/numa-node

done

> Consider hiding this field first and adding it if necessary

Keep the status for now, and then introduce a new protocol design when it is implemented later.